docs/pallas/tpu/matmul.md

```diff
@@ -0,0 +1,643 @@
+
+```{code-cell} ipython3
+:id: RjU3sSTUWzIk
+:outputId: 02b5793e-1ff3-41f4-ab45-4cf1393885ba
+
+import timeit
+
+def benchmark(f, ntrials: int = 100):
+  def run(*args, **kwargs):
+    # Compile function first
+    jax.block_until_ready(f(*args, **kwargs))
+    # Time function
+    result = timeit.timeit(lambda: jax.block_until_ready(f(*args, **kwargs)),
+                           number=ntrials)
+    time = result / ntrials
+    # print(f"Time: {time}")
+    return time
+  return run
+
+def analyze_matmul(m: int, k: int, n: int, dtype: np.dtype,
+                   mm_func):
+  x = jnp.ones((m, k), dtype=dtype)
+  y = jnp.ones((k, n), dtype=dtype)
+  time = benchmark(mm_func)(x, y)
+  print(f"----- {m} x {k} x {n} -----")
+  print("Matmul time: ", time)
+  mm_flops = matmul_flops(m, k, n) / time
+  print("Matmul FLOP/s: ", mm_flops)
+  print(f"FLOPs utilization: {mm_flops / v5e_bf16_flops * 100:.4f}%")
+  print()
+
+print("================bm=128, bk=128, bn=128===================")
+mm = functools.partial(matmul, bm=128, bk=128, bn=128)
+analyze_matmul(1024, 1024, 1024, jnp.bfloat16, mm)
+analyze_matmul(4096, 4096, 4096, jnp.bfloat16, mm)
+analyze_matmul(8192, 8192, 8192, jnp.bfloat16, mm)
+
+print("================bm=512, bk=1024, bn=1024===================")
+mm = functools.partial(matmul, bm=512, bk=1024, bn=1024)
+analyze_matmul(1024, 1024, 1024, jnp.bfloat16, mm)
+analyze_matmul(4096, 4096, 4096, jnp.bfloat16, mm)
+analyze_matmul(8192, 8192, 8192, jnp.bfloat16, mm)
+```
+
++++ {"id": "mg1GMqcVan70"}
+
+Bigger block sizes help a lot! We get pretty good utilization (80-90%) in the bigger matmuls, but the smallest matmul seems pretty hard to get good performance with.
+
+Let's compare this with XLA's matmuls. We don't expect Pallas to do better than XLA because XLA is *very* good at generating matmuls but hopefully we are close.
+
+```{code-cell} ipython3
+:id: OpU7I7BNXQYg
+:outputId: 28c2c3cf-759e-465c-f969-0e2c9607b8a5
+
+print("================ XLA matmul ===================")
+mm = jnp.matmul
+analyze_matmul(1024, 1024, 1024, jnp.bfloat16, mm)
+analyze_matmul(4096, 4096, 4096, jnp.bfloat16, mm)
+analyze_matmul(8192, 8192, 8192, jnp.bfloat16, mm)
+```
+
++++ {"id": "L-KUG3lha-jm"}
+
+Pallas, with some very basic tuning, gets pretty close to XLA's performance numbers! By trying out more block sizes, we should expect to close the gap entirely.
+
++++ {"id": "nbdHMJRObnZa"}
+
+## Templating the matrix multiplication
+
++++ {"id": "qSfcMwtDg7Vn"}
+
+Now that we have a basic matrix multiplication kernel, we can now try fusing operations into it.
+
+### Fused right-hand-side transpose
+
+A common first thing to do is to fuse a transpose (e.g. instead of doing `x @ y` we do `x @ y.T`). On TPUs, the MXU supports matmul routines with the right hand side transposed natively, so there is no additional cost to fusing in a transpose. We can use this routine with `jax.lax.dot_general`.
```

Comment:
I would speak less of MXU features, since those can vary between generations. And perhaps say that in Pallas we currently support transposed RHS, which can be lowered into more efficient hardware instructions compared to doing a transpose and matmul completely separately.